code/modules/library/library_computer.dm

```diff
@@ -0,0 +1,577 @@
+			if(orderedbook && print_cooldown <= world.time)
+				make_external_book(orderedbook)
+				print_cooldown = world.time + PRINTING_COOLDOWN
+		if("order_programmatic_book")
+			var/datum/programmatic_book/PB = GLOB.library_catalog.get_programmatic_book_by_id(params["bookid"])
+			if(PB && print_cooldown <= world.time)
+				make_programmatic_book(PB)
+				print_cooldown = world.time + PRINTING_COOLDOWN
+		//book author actions
+		if("delete_book")
+			if(params["bookid"])
+				var/datum/cachedbook/selectedbook = GLOB.library_catalog.get_book_by_id(params["bookid"])
+				if(!selectedbook)
+					playsound(src, 'sound/machines/synth_no.ogg', 15, TRUE)
+					atom_say("Deletion Failed!")
+				if(selectedbook.ckey != params["user_ckey"])
+					message_admins("[params["user_ckey"]] attempted to delete a book that wasn't theirs, this shouldn't happen, please investigate.")
+				if(GLOB.library_catalog.remove_book_by_id(params["bookid"])) //this doesn't need to be logged
+					playsound(loc, 'sound/machines/ping.ogg', 25, 0)
+					atom_say("Deletion Succesful!")
+				else
+					playsound(src, 'sound/machines/synth_no.ogg', 15, TRUE)
+					atom_say("Deletion Failed!")
+
+
+		//rating acts
+		if("set_rating")
+			if(params["rating_value"])
+				user_data.selected_rating = text2num(params["rating_value"])
+		if("rate_book")
+			if(GLOB.library_catalog.rate_book(params["user_ckey"], params["bookid"], user_data.selected_rating))
+				playsound(loc, 'sound/machines/ping.ogg', 25, 0)
+				atom_say("Rating Succesful!")
+			populate_booklist()
+		//Report Acts
+		if("submit_report")
+			if(GLOB.library_catalog.flag_book_by_id(params["user_ckey"], params["bookid"], selected_report))
+				playsound(loc, 'sound/machines/ping.ogg', 50, 0)
+				atom_say("Report Submitted!")
+			else
+				playsound(src, 'sound/machines/synth_no.ogg', 15, TRUE)
+				atom_say("Report Submission Failed!")
+		if("set_report")
+			selected_report =  text2num(params["report_type"])
+		//Book Uploader
+		if("toggle_upload_category")
+			if(text2num(params["category_id"]) in user_data.selected_book.categories)
+				user_data.selected_book.categories -= text2num(params["category_id"])
+				populate_booklist()
+			else
+				if(length(user_data.selected_book.categories) >= 3)
+					playsound(src, 'sound/machines/synth_no.ogg', 15, TRUE)
+					return
+				user_data.selected_book.categories += text2num(params["category_id"])
+				populate_booklist()
+		if("uploadbook")
+			if(GLOB.library_catalog.upload_book(params["user_ckey"], user_data.selected_book))
+				playsound(src, 'sound/machines/ping.ogg', 50, 0)
+				atom_say("Book Uploaded!")
+			else
+				playsound(src, 'sound/machines/synth_no.ogg', 15, TRUE)
+				atom_say("Book Upload Failed!")
+			num_pages = getmaxpages()
+		if("reportlost")
+			inventoryRemove(text2num(params["libraryid"]))
+			for(var/datum/borrowbook/book in checkouts)
+				if(book.libraryid == text2num(params["libraryid"]))
+					checkouts -= book
+	add_fingerprint(usr)
```

Comment:
You probably want this at the start of the proc. So you can use early returns in the code above

Reply:
done